### PR TITLE
[FIX] web_editor: update cols options after drag release

### PR DESCRIPTION
**Pull request by robinlej**
Since [commit 1], the columns count in the options is updated while dragging the width handle of a column. That behavior relies on asynchronous code which is not awaited.

This commit corrects it to be in line with the behavior of all other options: the count is updated after releasing the drag.

[commit 1]: https://github.com/odoo/odoo/commit/710d000f1872fd99b41d52ec3d6923756bba7cba

task-3576046

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -4689,6 +4689,9 @@ registry.sizing = SnippetOptionWidget.extend({
                 }, 0);
 
                 self.options.wysiwyg.odooEditor.automaticStepActive('resizing');
+                await new Promise(resolve => {
+                    self.trigger_up("snippet_option_update", { onSuccess: () => resolve() });
+                });
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4694]: What's that supposed to be doing? :grimacing: You make an event handler async, which has simply no effect other than being able to use `await` but you `await` something which is not considered by anything then. This code here would lead to the same result as `self.trigger_up("snippet_option_update", { onSuccess: () => {} });`

This should be better included in the main Mutex of the editor so that resizing + updating-the-ui-after-resizing are properly queue with other actions.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 1 above)
Comment (robinlej) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4694]: > What's that supposed to be doing?

I got tangled up I guess. But yes, in the end I wanted to use `self.trigger_up("snippet_option_update", { onSuccess: () => {} });`.

> This should be better included in the main Mutex of the editor so that resizing + updating-the-ui-after-resizing are properly queue with other actions.

Not sure what you mean here. Maybe it's because I don't fully understand how the actions and the mutex interact. What about resizing should be better included in the mutex?

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4694]: @robinlej Everything in the editor options is queued in the main mutex at the SnippetsMenu level: if you click on 3 options A, B, and C very quickly, what should happen is A, updateUI, B, updateUI, C, updateUI, regardless of the (asynchronous) time it takes for the editor to perform each action and each UI update. The goal being to have a consistent behavior (no race condition with synchronous action being done at random places in between asynchronous ones) and have the UI update... being done when it can be updated.

Here, this apparently works by chance: we have an option (resizing columns) which does not go through any of the `_select` logic / `_updateUI` logic etc. Simply triggering `snippet_option_update` at the right place is a functional improvement but it's probably a new source of race condition: you ask for an UI update "immediately" (without waiting for the current queue to be dealt with) and you don't wait for the UI update to be done, meaning any other option can be performed "meanwhile" the UI update.

What should be done here is probably something like the drag & drop: locking the mutex when starting the drag/resizing, unlocking once you are done resizing, then ask for an update UI which has to be added in the queue as well. Hint: `snippet_edition_request` ?

**Comment by robinlej** [comment]
@qsm-odoo I think it should be good.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -4518,6 +4518,7 @@ registry.sizing = SnippetOptionWidget.extend({
             ev.preventDefault();
             self.options.wysiwyg.odooEditor.automaticStepUnactive('resizing');
             isMobile = weUtils.isMobileView(self.$target[0]);
+            let resizeResolve;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4521]: Why do you need to define this this far from the initialization? Can't you declare this right above the first mutex lock?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -4689,6 +4701,9 @@ registry.sizing = SnippetOptionWidget.extend({
                 }, 0);
 
                 self.options.wysiwyg.odooEditor.automaticStepActive('resizing');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4703]: I am surprised that this is not called if we reach the `return` just above... is that right? In any case I think it would be better to free the mutex right after this (and if it happens that this should indeed be after the `return`, you would obviously need to free the mutex before the return too).

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -4689,6 +4701,9 @@ registry.sizing = SnippetOptionWidget.extend({
...
+                self.trigger_up("snippet_edition_request", { exec: () => {
+                    self.trigger_up("snippet_option_update", { onSuccess: () => {} });
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4705]: The whole point of using `snippet_edition_request` is that you can now properly wait for the UI update, giving something as the `onSuccess` value :wink:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 5]
```diff
@@ -4571,6 +4572,14 @@ registry.sizing = SnippetOptionWidget.extend({
                 return;
             }
 
+            // Locking the mutex during the resize. Started here to avoid
+            // empty returns.
+            const prom = new Promise(resolve => resizeResolve = () => resolve());
+            self.trigger_up("snippet_edition_request", { exec: () => {
+                self.trigger_up("disable_loading_effect");
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4579]: Are not you supposed to call `enable_loading_effect` at some point if you use that?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4703]: It's not probably not required. But from a consistency point of view, I would say "Ok the action is finished" (= freeing the edition mutex) at the very end.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 6]
```diff
@@ -4557,6 +4557,7 @@ registry.sizing = SnippetOptionWidget.extend({
             const isGrid = Object.keys(resizeValues).length === 4;
             const isGridHandle = $handle[0].classList.contains('o_grid_handle');
             if (isGrid && !isGridHandle || !isGrid && isGridHandle) {
+                self.options.wysiwyg.odooEditor.automaticStepActive("resizing");
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4560]: @robinlej Let's not mix fixes. I don't think this fix is related in any way to the mutex lock. And actually, while I said that the mutex lock can target 17.0 because "risky", this can probably target older versions and deserve its own commit explanation :wink:

**Comment by robinlej** [comment]
@qsm-odoo Commit updated.
I extracted the `automaticStepActive()` calls here: https://github.com/odoo/odoo/pull/145623

(I'll probably ping you once more on Monday. ;)

**Comment by robinlej** [comment]
@qsm-odoo As I said: pinging you on Monday, it's ready!

**Comment by qsm-odoo** [comment]
Note to myself, waiting for https://github.com/odoo/odoo/pull/145623 to be merged and forward-ported.
